src/server/routes/entity/process-unified-form.ts

```diff
@@ -242,116 +173,39 @@ export function handleCreateMultipleEntities(
-				isMerge: false
-			}).save(null, {transacting});
-			const additionalProps = _.pick(entityForm, additionalEntityProps[_.camelCase(entityType)]);
-			const changedProps = await getChangedProps(
-				orm, transacting, isNew, currentEntity, entityForm, entityType,
-				newRevision, additionalProps
-			);
-			// If there are no differences, bail
-			if (_.isEmpty(changedProps)) {
-				savedMainEntities[entityKey] = currentEntity;
-				_.set(savedMainEntities[entityKey], ['relationshipSet', 'id'], savedMainEntities[entityKey].relationshipSetId);
-				// remove the revision if no changes made
-				await newRevision.destroy({transacting});
-				return;
-			}
-			const mainEntity = await fetchOrCreateMainEntity(
-				orm, transacting, isNew, currentEntity.bbid, entityType
-			);
-			mainEntity.shouldInsert = isNew;
-
-			// set changed attributes on main entity
-			_.forOwn(changedProps, (value, key) => mainEntity.set(key, value));
-			const savedMainEntity = await saveEntitiesAndFinishRevision(
-				orm, transacting, isNew, newRevision, mainEntity, [mainEntity],
-				editorJSON.id, entityForm.note
-			);
-
-			/* We need to load the aliases for search reindexing and refresh it*/
-			await savedMainEntity.load('aliasSet.aliases', {transacting});
-
-			/* New entities will lack some attributes like 'type' required for search indexing */
-			await savedMainEntity.refresh({transacting});
-
-			/* fetch and reindex EditionGroups that may have been created automatically by the ORM and not indexed */
-			if (savedMainEntity.get('type') === 'Edition') {
-				await indexAutoCreatedEditionGroup(orm, savedMainEntity, transacting);
-			}
-			bbidMap[entityKey] = savedMainEntity.get('bbid');
-			savedMainEntities[entityKey] = savedMainEntity.toJSON();
-			// getNextRelationshipSet expects relationshipSet.id to exist
-			if (!isNew) {
-				_.set(savedMainEntities[entityKey], ['relationshipSet', 'id'], savedMainEntities[entityKey].relationshipSetId);
-			}
-		}
-		try {
-			// bookshelf's transaction have issue with Promise.All, refer https://github.com/bookshelf/bookshelf/issues/1498 for more details
-			await Object.keys(body).reduce((promise, entityKey) => promise.then(() => processEntity(entityKey)), Promise.resolve());
 
-			// adding relationship on newly created entites
-			await Object.keys(allRelationships).reduce((promise, entityId) => promise.then(() => processRelationship(
-				allRelationships[entityId], savedMainEntities[entityId], bbidMap, editorJSON.id, orm, transacting
-			)), Promise.resolve());
-
-			return savedMainEntities;
-		}
-		catch (err) {
-			log.error(err);
-			throw err;
-		}
-	});
-	const achievementPromise = entityEditPromise.then(
-		(entitiesJSON:Record<string, any>) => {
-			const entitiesAchievementsPromise = [];
-			for (const entityJSON of Object.values(entitiesJSON)) {
-				entitiesAchievementsPromise.push(processAchievement(orm, editorJSON.id, entityJSON));
+	const {body}: {body: Record<string, any>} = req;
```

Comment:
OK, I took a bit of time to rewrite some of the code from promises to async/await, I always get confused when I mix the two styles :)

You'll notice I separate the entity processing and achievement processing a bit, and I am passing an async function (which ~= a Promise) to sendPromiseResult rather than executing the function.
SendPromiseResult will take care of `await`-ing the promise for us in a try-catch block.

Here's what I came up with for the whole function:

```
export function handleCreateMultipleEntities(
	req: PassportRequest,
	res: $Response
) {
	const {orm}: {orm?: any} = req.app.locals;
	const editorJSON = req.user;
	let currentEntity: {
		__isNew__: boolean | undefined,
		aliasSet: {id: number} | null | undefined,
		annotation: {id: number} | null | undefined,
		bbid: string,
		disambiguation: {id: number} | null | undefined,
		identifierSet: {id: number} | null | undefined,
		type: EntityTypeString
	} | null | undefined;


	const {body}: {body: Record<string, any>} = req;
	async function processEntity(entityKey:string) {
		const entityForm = body[entityKey];
		const entityType = _.upperFirst(entityForm.type);
		const deriveProps = _.pick(entityForm, additionalEntityProps[_.camelCase(entityType)]);
		const isNew = _.get(entityForm, '__isNew__', true);
		currentEntity = null;
		if (!isNew) {
			currentEntity = await orm.func.entity.getEntity(orm, entityType, entityForm.id, getEntityRelations(entityType as EntityTypeString));
			if (!currentEntity) {
				throw new FormSubmissionError('Entity with given id not found');
			}
		}
		const {bookshelf} = orm;
		const entityJSON = await bookshelf.transaction((transacting) =>
			processSingleEntity(entityForm, isNew ? null : currentEntity, null, entityType, orm, editorJSON, deriveProps, false, transacting));

		return entityJSON;
	}
	// create all entities
	async function processAllEntities() {
		// For each entity in the request body
		const processedEntities = await Object.keys(body).reduce((promise, entityKey) =>
			promise
				// submit the entity
				.then(() => processEntity(entityKey))
				// process potential achievements for this edit
				.then(entityJSON => processAchievement(orm, editorJSON.id, entityJSON))
		// Empty promise for initial value sent to .reduce
		, Promise.resolve());
		return processedEntities;
	}
	return handler.sendPromiseResult(
		res,
		processAllEntities,
		_bulkIndexEntities
	);
}
```

Full disclaimer, I haven't tried to run the code, so I'll leave that for you to try, and I'll hop on to debugging that failing test ;)